### PR TITLE
chore: Separate lerna version and publish

### PR DESCRIPTION
**Pull request by wenytang-ms**
…ain branch operation

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -112,15 +112,10 @@ jobs:
...
+      - name: publish npm packages to npmjs.org by conventional-commits
+        if: ${{ github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/main' && github.event.inputs.conventional-commits == 'yes'}}
         run: |
-          npx lerna publish --conventional-prerelease --preid=rc --yes
-
-      - name: release stable npm packages to npmjs.org by conventional-commits
-        if: ${{ github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/main' && github.event.inputs.conventional-commits == 'yes' && github.event.inputs.preid == 'stable'}}
-        run: |
-          npx lerna publish --conventional-graduate --yes
+          npx lerna publish from-package --yes
```
Comment (adashen) [inline .github/workflows/cd.yml:118]: Did we test whether it will release all packages? What is the manual step if we need update the version manually?

Comment (wenytang-ms) [reply .github/workflows/cd.yml:118]: yes, "from-package" will look inside the latest tag which lerna version put on the branch, and will publish the version tagged on the latest commits.
manually step is:
1. release manager fetch origin main, and run "npm run version-rc"
2. make sure the rc/stable version change, then input "yes" or "y"
3. push to the origin main branch, and trigger the CD pipeline, must choose "conventional-commits" as "yes"

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -98,30 +100,25 @@ jobs:
           git config --global user.email 'yiz@microsoft.com'
 
       - name: release alpha npm packages to npmjs.org
-        if: ${{ github.event_name == 'schedule' }}
+        if: ${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/dev' && github.event.inputs.preid == 'alpha')}}
         run: |
           npx lerna version --no-private prerelease --preid=alpha.$(git rev-parse --short HEAD) --no-push --allow-branch dev --yes
```
Comment (adashen) [inline .github/workflows/cd.yml:105]: add no tag

Comment (wenytang-ms) [reply .github/workflows/cd.yml:105]: seem with no-git-tag-version will product some uncommited files, then lerna publish will fail

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -3,21 +3,13 @@ name: CD
...
-      force:
-        description: "force release even if no update(set --force-publish if necessary)"
-        required: false
-        default: ""
-      conventional-commits:
-        description: "bump up release version by lerna conventional commits(yes or no)"
         required: true
+        default: "rc"
+      skip-rc-version:
+        description: "skip bump up rc version by lerna conventional commits(yes or no)"
```
Comment (qinezh) [inline .github/workflows/cd.yml:11]: what's the scenario to skip bump rc version in this CD pipeline?

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -98,30 +105,25 @@ jobs:
...
 
-      - name: release stable npm packages to npmjs.org
-        if: ${{ github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/main' && !startsWith(github.event.inputs.version, 'pre') && github.event.inputs.conventional-commits == 'no'}}
-        run: |
-          npx lerna publish --no-private --dist-tag=latest ${{ github.event.inputs.version }} --no-changelog --yes ${{ github.event.inputs.force }}
-
-      - name: release rc npm packages to npmjs.org by conventional-commits
-        if: ${{ github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/main' && github.event.inputs.conventional-commits == 'yes' && github.event.inputs.preid == 'rc'}}
+      - name: publish rc npm packages to npmjs.org
+        if: ${{ github.event_name == 'workflow_dispatch' && github.ref == 'refs/heads/main' && github.event.inputs.preid == 'rc' }}
```
Comment (qinezh) [inline .github/workflows/cd.yml:119]: do we need to check `skip-rc-version` here?

File: package.json [hunk 5]
```diff
@@ -5,6 +5,7 @@
         "bootstrap": "lerna bootstrap",
         "lint:staged": "lerna run lint:staged --stream",
         "lint": "lerna run lint --stream",
+        "manual-version": "lerna version --preid=rc --no-push --no-changelog --no-private",
         "version": "lerna bootstrap --ignore-scripts -- --package-lock-only --no-audit && git add packages/*/package-lock.json",
```
Comment (qinezh) [inline package.json:9]: how about replacing with the original "version" script to avoid confusion?

File: .github/workflows/cd.yml [hunk 6]
```diff
@@ -27,8 +19,23 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - name: Validate inputs
-        if: ${{ github.event_name == 'workflow_dispatch' && ((github.ref != 'refs/heads/dev' && github.ref != 'refs/heads/main') || (github.ref == 'refs/heads/dev' && (!startsWith(github.event.inputs.version, 'pre') || github.event.inputs.preid != 'alpha')) || (github.ref == 'refs/heads/main' && startsWith(github.event.inputs.version, 'pre') && github.event.inputs.preid != 'rc') || (github.ref == 'refs/heads/dev' && github.event.inputs.conventional-commits == 'yes') || (github.ref == 'refs/heads/main' && github.event.inputs.conventional-commits == 'yes' && !(github.event.inputs.preid == 'rc' || github.event.inputs.preid == 'stable'))) }}
+      - name: Validate CD branch
+        if: ${{ github.event_name == 'workflow_dispatch' && github.ref != 'refs/heads/main' && github.ref != 'refs/heads/dev' }}
+        run: |
+          exit 1
```
Comment (qinezh) [inline .github/workflows/cd.yml:25]: would it better to output error message that why the pipeline exits?

File: .github/workflows/cd.yml (hunk 3 above)
Comment (qinezh) [reply .github/workflows/cd.yml:11]: offline discussed, may be we could rename it like `enable-auto-versioning`?

Comment (wenytang-ms) [reply .github/workflows/cd.yml]: just rename it as "skip-version-rc" and modify the description
actually, this config work for rc version, so I think it's better add rc on it

File: .github/workflows/cd.yml (hunk 4 above)
Comment (wenytang-ms) [reply .github/workflows/cd.yml:123]: skip-rc-verison is optional, not required, so there is no need to check

File: package.json (hunk 5 above)
Comment (wenytang-ms) [reply package.json]: this 'version' is removed, already sync up wth @LongOddCode

File: .github/workflows/cd.yml (hunk 6 above)
Comment (wenytang-ms) [reply .github/workflows/cd.yml:26]: sure!

File: package.json [hunk 7]
```diff
@@ -6,7 +6,6 @@
         "lint:staged": "lerna run lint:staged --stream",
         "lint": "lerna run lint --stream",
         "manual-version": "lerna version --preid=rc --no-push --no-changelog --no-private",
```
Comment (qinezh) [inline package.json:8]: would it be better to update to `version:rc` as the script will only version with preid `rc`?

Comment (wenytang-ms) [reply package.json:8]: sure

**Review by qinezh — APPROVED** [review]

**Review by LongOddCode — APPROVED** [review]

File: .github/workflows/cd.yml [hunk 8]
```diff
@@ -62,7 +69,7 @@ jobs:
             ./*.zip
 
       - name: release templates' rc version to github
-        if: ${{ github.ref == 'refs/heads/main' && startsWith(github.event.inputs.version, 'pre') }}
+        if: ${{ github.ref == 'refs/heads/main' && github.event.inputs.preid == 'rc' }}
         uses: marvinpinto/action-automatic-releases@latest
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply .github/workflows/cd.yml:77]: sure, I will investigate this later.

File: .github/workflows/cd.yml [hunk 9]
```diff
@@ -62,7 +69,7 @@ jobs:
             ./*.zip
 
       - name: release templates' rc version to github
-        if: ${{ github.ref == 'refs/heads/main' && startsWith(github.event.inputs.version, 'pre') }}
+        if: ${{ github.ref == 'refs/heads/main' && github.event.inputs.preid == 'rc' }}
```
[the remark this answers is not in the document]
Comment (wenytang-ms) [reply .github/workflows/cd.yml:76]: sure, it's seems will fail, later I will work on this workaround

**Pull request merged**